### ISSUE TITLE
Disk usage metrics for containerd

### ISSUE THREAD

**Issue by ribbybibby** [google/cadvisor#2785]
When switching from docker to containerd as my container runtime in Kubernetes, I noticed that `container_fs_usage_bytes` metrics were no longer being exported for my containers.

It looks like disk usage metrics aren't implemented for containerd, as noted by this comment: https://github.com/google/cadvisor/blob/v0.38.6/container/containerd/handler.go#L164-L165.

Disk usage is a pretty important metric to monitor, so I think, if possible, this should be added.

**Comment by elcomtik** [comment]
I have experienced the same issue. 

> It looks like disk usage metrics aren't implemented for containerd, as noted by this comment: https://github.com/google/cadvisor/blob/v0.38.6/container/containerd/handler.go#L164-L165.

There is some conversation about these metrics https://github.com/containerd/containerd/issues/678. I suppose that contained provide this information.

**Comment by yyrdl** [comment]
PR was submited  . https://github.com/google/cadvisor/pull/2872

**Comment by jepio** [comment]
PR #2872 was closed, in favor of #2956, which was merged and subsequently reverted in #2964. The result is that these metrics are not available.

Is someone working on an alternative approach?

**Comment by sarbajitc** [comment]
Is there any timeline for a fix of this issue?

**Comment by baasumo** [comment]
+1 on looking for any update or timeline regarding this issue - these metrics are pretty important for observability and workload behavior.

**Comment by fernandesnikhil** [comment]
Adding onto this ticket since we're blocked on switching to the `containerd` CRI without these metrics. We have alerting around ephemeral file system usage that would break if cAdvisor doesn't collect these from `containerd`.

**Comment by snuggie12** [comment]
@bobbypage Do you have an update on this? Best I can follow is that there is a possibly-working version in the `containerd-cri` branch after #2966 was merged. However, it might be incomplete based on https://github.com/google/cadvisor/pull/2936#issuecomment-971619663?

Alternatively it seems like work has gone into not using cadvisor for container stats and k8s 1.23 has an alpha feature-gate which [uses the cri stats provider](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-gates) (`PodAndContainerStatsFromCRI`). Is the plan to put momentum into that instead? If so, do you know when it would go beta?

**Comment by brandond** [comment]
The `PodAndContainerStatsFromCRI` feature-gate does not seem to work either; at least with containerd 1.6.4 the stats are still missing.

**Comment by brandond** [comment]
It appears that this won't be addressed any time soon, as  [KEP-2371] (https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2371-cri-pod-container-stats/README.md) moves most of the stats collection out of cadvisor into the CRI interface. Is there an interim solution for users that need these stats?

**Comment by bobbypage** [comment]
The workaround for now is to use the `containerd-cri` branch which has a special patch to export containerd disk metrics. The following image can be used: `gcr.io/cadvisor/cadvisor:v0.45.0-containerd-cri`

**Comment by brandond** [comment]
Is that branch being actively maintained? Do you know if it still works normally with other runtimes?

**Comment by bobbypage** [comment]
> Is that branch being actively maintained? 
Yes, it is maintained we just pushed the latest v0.45.0 changes to this branch. The reason when need this separate branch is because to get the Disk usage metrics on containerd requires importing the CRI API into cAdvisor. However, we can't import the CRI API into cAdvisor because cAdvisor is imported by k8s and k8s itself includes the CRI API which results in a circular dependency. So the workaround for now is to have this separate branch which includes CRI API. (see https://github.com/google/cadvisor/pull/2872#discussion_r655783226 for that discussion).

> Do you know if it still works normally with other runtimes?

Yes, it will work with other runtimes as well, but if containerd is not used there is no benefit of using it.

**Comment by brandond** [comment]
Ah hmm. So I take it the circular dependency prohibits this branch from being embedded in the kubelet, and there's no easy path towards doing so? Running a standalone deployment of cadvisor isn't particularly palatable, as asking our users to retool their monitoring stacks to make use of that would be a non-trivial amount of work. I'm honestly surprised that we got this far into the dockershim depreciation with cadvisor missing feature parity for one of the most popular replacement runtimes.

**Comment by bobbypage** [comment]
@brandond are you referring that most folks are using the existing `/cadvisor/metrics` endpoint on kubelet? If that's the case, then yes, unfortunately we aren't able to bring back this patch into kubelet due to circular dependency issue. The KEP https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2371-cri-pod-container-stats/README.md aims to solve this issue long term.

**Comment by brandond** [comment]
@bobbypage do you know if there are any images published off that containerd-cri branch that we could use?

**Comment by snuggie12** [comment]
Would it be possible to list a config somewhere that only gathers the disk? GKE controls our normal cadvisor so running a minimal "container disk metrics only" daemonset seems like a simple work around.

**Comment by yvespp** [comment]
`kubectl get --raw "/api/v1/nodes/(node-name)/proxy/stats/summary"` (from kublet) gives infos for ephemeral-storage for each pod but sadly it's not available as a Prometheus metric...

**Comment by george-angel** [comment]
Refreshing my memory on this issue, I realised we didn't link to the exporter @ribbybibby written to address this issue: https://github.com/utilitywarehouse/kube-summary-exporter. We have been running it for nearly 2 years now.

**Comment by vasireddy99** [comment]
Hi, @bobbypage Is this [comment](https://github.com/google/cadvisor/issues/2785#issuecomment-1204837267) still valid, so shall we expect the similar [release tag](https://github.com/google/cadvisor/releases/tag/v0.45.0-containerd-cri) for [container 47.1](https://github.com/google/cadvisor/releases/tag/v0.47.1) as well ? and for the rest of the releases untill this is fixed in [enhancement/KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2371-cri-pod-container-stats/README.md#motivation). can you please confirm if you would recommend using this `containerd-cri` tags for the fix. seem like it is only workaround available. please correct me?

**Comment by sethAmazon** [comment]
Does containerd-cri work? I did a replace in my go mod and it still did not work. I see spec sets has file system to false. https://github.com/google/cadvisor/blob/containerd-cri/container/containerd/handler.go#L287-L289

**Comment by markrity** [comment]
Any updates on this ? containerd is the default and recommended runtime for GKE , but there is still no support for `kubernetes_filesystem_usage` ?

**Comment by sidewinder12s** [comment]
It appears at least on `containerd://1.6.6` and the `v0.45.0-containerd-cri` tag, the `container_fs_* metrics are also just wrong. 

`container_fs_usage_bytes` at least seems to be reporting the root device free space for every pod on the node as opposed to each containers usage. Does anyone have a reference config to use for containerd + that `containerd-cri` tag?

**Comment by dragosrosculete** [comment]
Any hope for this to be implemented soon ?

**Comment by smileusd** [comment]
Any updates?

**Comment by mikkeloscar** [comment]
I tried to rebase (https://github.com/google/cadvisor/tree/containerd-cri) on v0.48.0 (and v0.47.1) in both cases the resource usage blows up: 🙁

![image](https://github.com/google/cadvisor/assets/128566/bb6e1fa8-dc0d-4234-8162-61098f670985)

I do see values for the metrics, but didn't validate that they are correct as is reported not to be in other comments.

**Comment by brandond** [comment]
I doubt this is going to be fixed, given the work in progress to move stats into the CRI API and remove cadvisor from the kubelet: 

https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2371-cri-pod-container-stats/README.md#cadvisor-less-cri-full-container-and-pod-stats 
> enhance the CRI API with enough metrics to be able to supplement the pod and container fields in the summary API directly from CRI.
> enhance the CRI implementations to broadcast the required metrics to fulfill the pod and container fields in the /metrics/cadvisor endpoint.

**Comment by tripitakav** [comment]
Is Have Any updates?

**Comment by ning1875** [comment]
use the crictl tool can get container fs usage ,eg
```shell
 crictl stats
CONTAINER           CPU %               MEM                 DISK                INODES
0674440a33dbd       0.00                1.438MB             102.4kB             24
2e2f101e7ce72       0.06                62.43MB             114.7kB             29
37ed67b1e33cf       1.58                346.1MB             110.6MB             41

```
and the data show in DISK row  come from [this code ](https://github.com/kubernetes-sigs/cri-tools/blob/master/cmd/crictl/container_stats.go#L194)
crictl called cadvisor ListContainerStats api with grpc 
and the response has a key named WritableLayer  mean  container fs usage
```shell
type ContainerStats struct {
	// Information of the container.
	Attributes *ContainerAttributes `protobuf:"bytes,1,opt,name=attributes,proto3" json:"attributes,omitempty"`
	// CPU usage gathered from the container.
	Cpu *CpuUsage `protobuf:"bytes,2,opt,name=cpu,proto3" json:"cpu,omitempty"`
	// Memory usage gathered from the container.
	Memory *MemoryUsage `protobuf:"bytes,3,opt,name=memory,proto3" json:"memory,omitempty"`
	// Usage of the writable layer.
	WritableLayer *FilesystemUsage `protobuf:"bytes,4,opt,name=writable_layer,json=writableLayer,proto3" json:"writable_layer,omitempty"`
	// Swap usage gathered from the container.
	Swap                 *SwapUsage `protobuf:"bytes,5,opt,name=swap,proto3" json:"swap,omitempty"`
	XXX_NoUnkeyedLiteral struct{}   `json:"-"`
	XXX_sizecache        int32      `json:"-"`
}

```
so containerd has ability to get container fs usage ，but why cadvisor not call this ListContainerStats api？

**Comment by brandond** [comment]
I believe that is gated on the `PodAndContainerStatsFromCRI` FeatureGate, which is still alpha?

**Comment by wolgod** [comment]
+1 Is there any solution now？

**Comment by changhyuni** [comment]
What happened to the usage metric?
Where should I check?

**Comment by robini** [comment]
Is there any solution to find disk usage metrics for containerd via prometheus ?

**Comment by phuslu** [comment]
Same issue in our side, I work out a simple fix in https://github.com/google/cadvisor/pull/3569

**Comment by mindw** [comment]
@robini - quoting @george-angel :
> Refreshing my memory on this issue, I realised we didn't link to the exporter @ribbybibby written to address this: https://github.com/utilitywarehouse/kube-summary-exporter. We have been running it for nearly 2 years now.

**Comment by sidewinder12s** [comment]
Yep also been running that for about a year.

**Comment by luarx** [comment]
Is there any timeline for a fix of this issue? 🙏 
We need to monitor the storage used by container (`container_fs_usage_bytes`) and not only the ephemeral storage because we are not always using ephemeral storages

**Comment by vishiy** [comment]
Hi all - when can we expect fix for this ?

**Comment by dims** [comment]
Closing this issue as it has been open since before January 1, 2024. If this is still relevant, please feel free to reopen or create a new issue.

**Comment by till** [comment]
@dims I understand project bandwidth is limited, why not respond to people on this issue, or work on the latest PR before closing this?

**Comment by dims** [comment]
I am happy to reopen if someone wants to pick this up and work on it.

**Comment by george-angel** [comment]
> I am happy to reopen if someone wants to pick this up and work on it.

https://github.com/google/cadvisor/pull/3502 ?
